### DISCUSSION TITLE
ApiMan plugin - validate response data

### DISCUSSION THREAD

**Discussion by radlik01** [apiman/apiman#2113] [Help] [answered]
Dear all,

I would like to ask you whether it is possible to validate response data and change the response code depending on data validation.

something like

```  
public IReadWriteStream<ApiResponse> getResponseDataHandler(ApiResponse response, IPolicyContext context, Object policyConfiguration) {
    sLog.infof("getResponseDataHandler   begin");
    final IBufferFactoryComponent bufferFactory = context.getComponent(IBufferFactoryComponent.class);
    return new AbstractStream<ApiResponse>() {

      @Override
      public ApiResponse getHead() {
        sLog.infof("getResponseDataHandler   getHead");
        return response;
      }

      @Override

      public void write(IApimanBuffer chunk) {
        sLog.infof("getResponseDataHandler   chunk : %s", chunk.toString());
        responseData = chunk.toString();
        super.write(chunk);
      }


      @Override

      public void end() {
        sLog.infof("getResponseDataHandler   END");
        if (some data condition) {
          response.setCode(404);
        }
        super.end();

      }

      @Override
      protected void handleHead(ApiResponse head) {
        sLog.infof("getResponseDataHandler   READ");
      }

    };

  }
```

but this is not changed the response code. If I changed the response code before returning AbstractStream<ApiResponse>() than the code was changed successfully.

Thank you

Regards

Jan

**Comment by msavy** [comment, marked as answer]
Hi Jan,

Oh, that's an interesting one.

Some possibilities worth exploring:

Apiman-based:
*  Enable stateful request payload inspection 
* I don't recall whether this also works for the response, but you may be able to get the ordering you are looking for (i.e. can delay the response for long enough to parse the body). Let us know.
![image](https://user-images.githubusercontent.com/423513/173605246-6cf0ce34-7e46-4a52-a2c1-72f2ccffc123.png)

Exception/error-based
* Throw an error/exception before closing the connection, this IO error should propagate to the client -- albeit the return code will not change.

Body
* Return a payload that is something like JSON and have a embedded status code/message, use that instead of the HTTP response code.

Trailer
* Another possible option would be to set an HTTP trailer. Unfortunately this is not supported in Apiman at the moment, as it's quite a rare thing that very few people actually use. It's essentially like an HTTP header, but at the end of the request. AFAIK, trailers are only supported with `chunked` encoding.

**Reply by msavy** [reply]
p.s. you probably want to extend `AbstractMappedDataPolicy`

**Reply by msavy** [reply]
@radlik01 have you progressed with this?

**Comment by radlik01** [comment]
I think the only way is to raise IO Exception. This is not 100% what we want but it is work.
But would be very nice to return the different return code with payload like the classic REST API did
